clang/lib/CIR/CodeGen/CIRGenDecl.cpp

```diff
@@ -322,18 +333,277 @@ void CIRGenFunction::buildVarDecl(const VarDecl &D) {
+    llvm_unreachable("OpenCL & CUDA are NYI");
+  else
+    Init = builder.getZeroInitAttr(getTypes().ConvertType(Ty));
+
+  mlir::cir::GlobalOp GV = builder.createVersionedGlobal(
+      getModule(), getLoc(D.getLocation()), Name, LTy, false, Linkage);
+  // TODO(cir): infer visibility from linkage in global op builder.
+  GV.setVisibility(getMLIRVisibilityFromCIRLinkage(Linkage));
+  GV.setInitialValueAttr(Init);
+  GV.setAlignment(getASTContext().getDeclAlign(&D).getAsAlign().value());
+
+  if (supportsCOMDAT() && GV.isWeakForLinker())
+    llvm_unreachable("COMDAT globals are NYI");
+
+  if (D.getTLSKind())
+    llvm_unreachable("TLS mode is NYI");
+
+  setGVProperties(GV, &D);
+
+  // Make sure the result is of the correct type.
+  assert(!UnimplementedFeature::addressSpace());
+
+  // Ensure that the static local gets initialized by making sure the parent
+  // function gets emitted eventually.
+  const Decl *DC = cast<Decl>(D.getDeclContext());
+
+  // We can't name blocks or captured statements directly, so try to emit their
+  // parents.
+  if (isa<BlockDecl>(DC) || isa<CapturedDecl>(DC)) {
+    DC = DC->getNonClosureContext();
+    // FIXME: Ensure that global blocks get emitted.
+    if (!DC)
+      llvm_unreachable("address space is NYI");
+  }
+
+  GlobalDecl GD;
+  if (const auto *CD = dyn_cast<CXXConstructorDecl>(DC))
+    llvm_unreachable("C++ constructors static var context is NYI");
+  else if (const auto *DD = dyn_cast<CXXDestructorDecl>(DC))
+    llvm_unreachable("C++ destructors static var context is NYI");
+  else if (const auto *FD = dyn_cast<FunctionDecl>(DC))
+    GD = GlobalDecl(FD);
+  else {
+    // Don't do anything for Obj-C method decls or global closures. We should
+    // never defer them.
+    assert(isa<ObjCMethodDecl>(DC) && "unexpected parent code decl");
+  }
+  if (GD.getDecl() && UnimplementedFeature::openMP()) {
+    // Disable emission of the parent function for the OpenMP device codegen.
+    llvm_unreachable("OpenMP is NYI");
+  }
+
+  return GV;
+}
+
+/// Add the initializer for 'D' to the global variable that has already been
+/// created for it. If the initializer has a different type than GV does, this
+/// may free GV and return a different one. Otherwise it just returns GV.
+mlir::cir::GlobalOp
+CIRGenFunction::addInitializerToStaticVarDecl(const VarDecl &D,
+                                              mlir::cir::GlobalOp GV) {
+  ConstantEmitter emitter(*this);
+  mlir::TypedAttr Init =
+      emitter.tryEmitForInitializer(D).dyn_cast<mlir::TypedAttr>();
+  assert(Init && "Expected typed attribute");
+
+  // If constant emission failed, then this should be a C++ static
+  // initializer.
+  if (!Init) {
+    if (!getLangOpts().CPlusPlus)
+      CGM.ErrorUnsupported(D.getInit(), "constant l-value expression");
+    else if (D.hasFlexibleArrayInit(getContext()))
+      CGM.ErrorUnsupported(D.getInit(), "flexible array initializer");
+    else if (HaveInsertPoint()) {
```

Comment:
Do this actually get hit? Differently from LLVM codegen, we don't clear insertion points to signal behavior to other parts of codegen - this happens elsewhere and IMO is a bad design from LLVM Codegen. Suggestion: perhaps we do instead:
```
else {
    // Since we have a static initializer, this global variable can't
    // be constant.
    GV.setConstant(false);
    llvm_unreachable("C++ guarded init it NYI");
}
```
If this is not possible because of other things I'm missing, it's fine to leave as is for now, but add a comment in the lines of what I'm saying here.